### ISSUE TITLE
Improve ranking system

### ISSUE THREAD

**Issue by MindLaborDev** [anuraghazra/github-readme-stats#455]
**Describe the bug**
I was interested in how the rank was calculated and wanted to know how to rank up. I found out that you use a function called normalcdf to normalize the calculated score. Its basically impossible to get a rank below A+ since normalcdf(0, 231, 7) * 100 returns 51.55 which is in the A+ area. 

The code can be found [here](https://github.com/anuraghazra/github-readme-stats/blob/master/src/calculateRank.js).

#### TODOS (Added by @rickstaa)

- [x] Decide on a new grading system, #2265.
    - [x] Create PR that adds this new grading system.
- [x] Merge https://github.com/anuraghazra/github-readme-stats/pull/2116.
- [x] Merge https://github.com/anuraghazra/github-readme-stats/pull/1186.
- [x] Create PR that shows the top user percentage on hover (see https://github.com/anuraghazra/github-readme-stats/issues/2157).

**Comment by nrchandan** [comment]
Created a PR that demonstrates this: #459

**Comment by junghans** [comment]
0 contributions still A+, e.g. https://github-readme-stats.vercel.app/api?username=marlinofdoom

**Comment by MarvinJWendt** [comment]
## This is still very relevant
![image](https://user-images.githubusercontent.com/31022056/97818827-15736e00-1ca5-11eb-9cf2-b2effc5a10f5.png)
![image](https://user-images.githubusercontent.com/31022056/97818819-0b516f80-1ca5-11eb-9042-c64d6ae15284.png)

/cc @anuraghazra

**Comment by SuperFola** [comment]
Commenting to get this issue fixed

/cc @anuraghazra

**Comment by SuperFola** [comment]
This is still an issue

**Comment by magneticflux-** [comment]
I think it might be broken in more ways than one.
- The implementation doesn't seem to work right: https://stackoverflow.com/questions/5259421/cumulative-distribution-function-in-javascript/5263759#comment114047028_5263759
- The CDF used is for the normal distribution, which I doubt GitHub stats adheres to. Perhaps a power-law distribution function would be a better fit? https://en.wikipedia.org/wiki/Power_law#Power-law_probability_distributions

**Comment by magneticflux-** [comment]
Still an issue

**Comment by anuraghazra** [comment]
> I think it might be broken in more ways than one.
> 
> * The implementation doesn't seem to work right: https://stackoverflow.com/questions/5259421/cumulative-distribution-function-in-javascript/5263759#comment114047028_5263759
> * The CDF used is for the normal distribution, which I doubt GitHub stats adhere to. Perhaps a power-law distribution function would be a better fit? https://en.wikipedia.org/wiki/Power_law#Power-law_probability_distributions
> 
> Here's another example of the weird ranking behavior: [#330 (comment)](https://github.com/anuraghazra/github-readme-stats/issues/330#issuecomment-689773382)

I agree. rank evaluation is not correct and too biased. 
Sorry but I have no good ideas on how to make it less inaccurate and improve it (i suck at maths) 
Would love to hear any opinions and possible implementation details. 

cc @magneticflux-

**Comment by CrafterKolyan** [comment]
@anuraghazra I've made an implementation of one of my ideas in #960 with an explanation inside PR

**Comment by rickstaa** [comment]
Just placing this here so we don't forget. As pointed out by #2116, there also is a bug in the current rank implementation that skews the scores.

**Comment by LucBerge** [comment]
I'm I the only one thinking that the ranking system is weird ? #2265

**Comment by ZjzMisaka** [comment]
> > I think it might be broken in more ways than one.
> > 
> > * The implementation doesn't seem to work right: https://stackoverflow.com/questions/5259421/cumulative-distribution-function-in-javascript/5263759#comment114047028_5263759
> > * The CDF used is for the normal distribution, which I doubt GitHub stats adhere to. Perhaps a power-law distribution function would be a better fit? https://en.wikipedia.org/wiki/Power_law#Power-law_probability_distributions
> > 
> > Here's another example of the weird ranking behavior: [#330 (comment)](https://github.com/anuraghazra/github-readme-stats/issues/330#issuecomment-689773382)
> 
> I agree. rank evaluation is not correct and too biased. Sorry but I have no good ideas on how to make it less inaccurate and improve it (i suck at maths) Would love to hear any opinions and possible implementation details.
> 
> cc @magneticflux-

Hi,
To be honest I don't care if my ranking is lowered after the ranking system is updated, but I don't like being stuck at A+ no matter how hard I try...
Another year has passed and I see a lot of people coming up with great ideas, maybe it's time to update the ranking system.

**Comment by rickstaa** [comment]
#### TODOS

- [ ] Decide on a new grading system.
- [ ] Create PR that uses a more commonly used grading system (i.e. A-F, A+ to F or % based).
- [ ] Merge https://github.com/anuraghazra/github-readme-stats/pull/2116.
- [ ] Merge https://github.com/anuraghazra/github-readme-stats/pull/1186.
- [ ] Create PR that shows the top user percentage on hover (see https://github.com/anuraghazra/github-readme-stats/issues/2157).

**Comment by ZjzMisaka** [comment]
Hi @rickstaa, 
I understand your caution in modifying the ranking system, but I believe that making changes is always good. 
Can some temporary modifications based on existing ranking system be made before receiving a response from @anuraghazra, at least to change the situation where the ranking is stuck in A+?
I think ranking can reflect the changes made by user from the initial state, rather than the gap between users.
Therefore, I don't care if my ranking will change to F after the ranking system changes, I prefer to see my progress dynamically.
After all, modifying a ranking system cannot satisfy everyone. If you insist on giving everyone a "beautiful" ranking, the system will ultimately become non referential.
Thanks.

**Comment by rickstaa** [comment]
> Hi @rickstaa, I understand your caution in modifying the ranking system, but I believe that making changes is always good. Can some temporary modifications based on existing ranking system be made before receiving a response from @anuraghazra, at least to change the situation where the ranking is stuck in A+? I think ranking can reflect the changes made by user from the initial state, rather than the gap between users. Therefore, I don't care if my ranking will change to F after the ranking system changes, I prefer to see my progress dynamically. After all, modifying a ranking system cannot satisfy everyone. If you insist on giving everyone a "beautiful" ranking, the system will ultimately become non referential. Thanks.

I agree with you. I already started working on this last week. Not only that, but I will merge https://github.com/anuraghazra/github-readme-stats/pull/1186#issuecomment-1537116588 as soon as @francois-rozet has resolved some small issues I had with that pull requests 🚀. @anuraghazra can always change it to his liking later.

**Comment by rickstaa** [comment]
Just a heads-up. I just merged https://github.com/anuraghazra/github-readme-stats/pull/1186 which, in my opinion, is a great improvement compared to the old ranking. You are welcome to open pull requests if you think the weights can be further improved 👍🏻. There is also a proposal open to change the rank levels. You are welcome to put your ideas in htps://github.com/anuraghazra/github-readme-stats/issues/2265.

**Comment by rickstaa** [comment]
Closing this as completed.

**Comment by rickstaa** [comment]
Re-opened so that people can provide feedback on the new ranking system and create improvements.

**Comment by lorezyra** [comment]
Would be good to update the readme for this project to include an explanation on the ranking system and meaning of the letters.

**Comment by rickstaa** [comment]
> Would be good to update the readme for this project to include an explanation on the ranking system and meaning of the letters.

You're right! The documentation will be updated when #2749 is merged 👍🏻. You are welcome to suggest any improvements.

**Comment by objektwerks** [comment]
So with the new ranking algo, a former ranking of 'S' for 12K plus annualized commits earns a modest ranking of 'B'. My ego may never recover. ;)

**Comment by rickstaa** [comment]
> So with the new ranking algo, a former ranking of 'S' for 12K plus annualized commits earns a modest ranking of 'B'. My ego may never recover. ;)

I'm afraid you are an outlier 😮. You made as many commits in one year as I did in my whole GitHub career 🔥. In my opinion, PRs are preferred over small commits since it improves code collaboration 🤔. I do understand that this is opinioned and that in some cases it is better to work with commits. Can you maybe check if @francois-rozet new ranking mechanism improves your rank (see https://github.com/anuraghazra/github-readme-stats/pull/2762)?

**Comment by rickstaa** [comment]
@objektwerks I checked and #2762 will improve the behaviour. Also by making some pull requests you can increase your rank rapidly. 

#### Old ranks

> Available ranks are S+ (top 2.5%), S (top 10%), A+ (top 25%), A (top 50%), B+ (top 75%) and B (everyone). This ranking scheme is based on the [Japanese Academic Grading system](https://en.wikipedia.org/wiki/Academic_grading_in_Japan). It is calculated as a weighted sum of ranks for each statistic (number of commits, pull requests, issues, stars and followers) based on the cumulative distribution function of an [exponential distribution](https://wikipedia.org/wiki/exponential_distribution). The implementation can be investigated at [src/calculateRank.js](./src/calculateRank.js). The circle around the rank shows the rank completion percentage compared to the maximum rank.

[![Anurag's GitHub stats](https://github-readme-stats.vercel.app/api?username=objektwerks)](https://github.com/anuraghazra/github-readme-stats)

#### Ranks (after #2762)

> S, A+, A, A-, B+, B, B-, C+ and C

[![Anurag's GitHub stats](https://github-readme-stats-git-rank-rickstaa.vercel.app/api?username=objektwerks)](https://github.com/anuraghazra/github-readme-stats)

**Comment by objektwerks** [comment]
@rickstaa Thanks for the explanation and tip! You all might consider 2 rankings -- 1) **pr**-driven and 2) **commit**-driven. Just a thought. :) Cheers!

**Comment by francois-rozet** [comment]
> @rickstaa Thanks for the explanation and tip! You all might consider 2 rankings -- 1) **pr**-driven and 2) **commit**-driven. Just a thought. :) Cheers!

Hello, I think making a lot of commits should not be a goal. Having 10000 commits (named "fix") that change one or two lines, is a bad development practice. PRs are much better indicators of meaningful contributions.

**Comment by objektwerks** [comment]
@francois-rozet Everyone has different development styles. Hence the multi-style ranking suggestion. :)

**Comment by rickstaa** [comment]
> @francois-rozet Everyone has different development styles. Hence the multi-style ranking suggestion. :)

I agree with @francois-rozet on this one. Although everybody has a different development style, from experience I know that it can hurt maintainability working projects with other people 😅. Nevertheless, since this is a community-based project, feel free to open a feature request and if enough people upvote your feature I will take a look 👍🏻 (see https://github.com/anuraghazra/github-readme-stats/issues/1935 for the highest-ranked features and bugs).

**Comment by rickstaa** [comment]
> @francois-rozet Everyone has different development styles. Hence the multi-style ranking suggestion. :)

> @francois-rozet Everyone has different development styles. Hence the multi-style ranking suggestion. :)

I agree with @francois-rozet on this one 😅. Although everybody has a different development style, from experience I noticed that these small commits hurt maintainability when working on projects with other people. Nevertheless, since this is an OS (community based) project, if enough people support your idea we might look at it (see https://github.com/anuraghazra/github-readme-stats/issues/1935 for the most popular feature requests) 👍🏻.

**Comment by objektwerks** [comment]
@rickstaa

I work alone on multiple projects. And small units of work make sense for my workflow. Not everyone is an OSS warrior. ;)

On a commercial or open source project, with several contributors, developers ideally work in a branch and submit PRs as required. So commit size or frequency is irrelevant. It's the PR process that impacts a project.

Ok, I've said way too much. :) Thanks for the conversation! And all the best with your project. Cheers!

**Comment by rickstaa** [comment]
> @rickstaa
> 
> I work alone on multiple projects. And small units of work make sense for my workflow. Not everyone is an OSS warrior. ;)
> 
> On a commercial or open source project, with several contributors, developers ideally work in a branch and submit PRs as required. So commit size or frequency is irrelevant. It's the PR process that impacts a project.
> 
> Ok, I've said way too much. :) Thanks for the conversation! And all the best with your project. Cheers!

No, you exactly hit the nail on the head. For personal projects, it doesn't really matter and in commercial projects or big OS projects, mostly PR with squashed commits are used 👍🏻. It's not that I have anything against people that add many commits to the master branch. It is only that me and the other collaborators/contributors are trying to promote PR-based contributions so that people will have less friction onboarding to bigger projects.

Having that said, this is a community project, so all input is valued equally. Furthermore, it's easy to increase the weights for the commits on your own fork 👍🏻. I think these ranks should be used as a fun motivational tool, not as a single source of truth. If you get more motivated to strengthen your coding skills and OS contributions by having the rank rise yearly by the number of commits, that's also great ❤️‍🔥.

**Comment by rickstaa** [comment]
This feature was completed in https://github.com/anuraghazra/github-readme-stats/pull/2762. You are welcome to open a new issue or pull request if you guys have more ideas to improve the ranking. One idea, for example, I think would be great to have is the following:

>Maybe we can give the user the ability to set the `rank_type` to `number` or `letter`? We could use the inverse of the number that is used to calculate the rank 🤔.

_Originally posted by @rickstaa in https://github.com/anuraghazra/github-readme-stats/issues/2265#issuecomment-1569571714_

**Comment by rickstaa** [comment]
I just created https://github.com/anuraghazra/github-readme-stats/pull/2858, which, when merged, gives users the ability to show the rank progress instead.

```md
![Anurag's GitHub stats](https://github-readme-stats-git-addprogressrankicon-rickstaa.vercel.app/api?username=anuraghazra\&rank_icon=progress)
```

![Anurag's GitHub stats](https://github-readme-stats-git-addprogressrankicon-rickstaa.vercel.app/api?username=anuraghazra\&rank_icon=progress)

**Comment by rickstaa** [comment]
Additionally, I now created #2859, which gives users the ability to show the `percentile`:

```md
[![Anurag's GitHub stats](https://github-readme-stats-git-addpercentilerankicon-rickstaa.vercel.app/api?username=mindlabordev&rank_icon=percentile)](https://github.com/anuraghazra/github-readme-stats)
```

[![Anurag's GitHub stats](https://github-readme-stats-git-addpercentilerankicon-rickstaa.vercel.app/api?username=mindlabordev&rank_icon=percentile&random=12345)](https://github.com/anuraghazra/github-readme-stats)

**Comment by rickstaa** [comment]
@silverwind, we decided to replace progress with percentile since it is unclear to users what progress means (see https://github.com/anuraghazra/github-readme-stats/pull/2859#pullrequestreview-1487346398). Therefore, changing progress to percentile in your README would be best.
